### ISSUE TITLE
Expected both objects to be of Result Type: Select

### ISSUE THREAD

**Issue by dneto0** [google/clspv#71]
This is forked from comment https://github.com/google/clspv/issues/2#issuecomment-333941818

The problem occurs when doing a somewhat simple selection between two loads from Private storage class (or __constant in OpenCL C).   The inst-combine optimization changes
-  a selection between two loads
into 
- a selection between two pointers, followed by a load from that pointer.

In the current compilation scheme the two pointers are in Private storage class, and the compiler incorrectly has the result of the selection being a pointer to StorageBuffer storage class.

Case is as below:
 
-------------
This is the symptom of compiling code like:

```
__constant float kFirst[3] = {1.0f, 2.0f, 3.0f};
__constant float kSecond[3] = {10.0f, 11.0f, 12.0f};

kernel void foo(global float*A, int c, int i) {
 *A = c==0 ? kFirst[i] : kSecond[i];
}
```

Produces this kind of code:
```
         %33 = OpVariable %_ptr_Private__arr_float_uint_3 Private %23
         %34 = OpVariable %_ptr_Private__arr_float_uint_3 Private %27
         %35 = OpVariable %_ptr_StorageBuffer__struct_4 StorageBuffer
         %36 = OpVariable %_ptr_StorageBuffer__struct_7 StorageBuffer
         %37 = OpVariable %_ptr_StorageBuffer__struct_7 StorageBuffer
         %38 = OpFunction %void None %11
         %39 = OpLabel
         %40 = OpAccessChain %_ptr_StorageBuffer_float %35 %uint_0 %uint_0
         %41 = OpAccessChain %_ptr_StorageBuffer_uint %36 %uint_0
         %42 = OpLoad %uint %41
         %43 = OpAccessChain %_ptr_StorageBuffer_uint %37 %uint_0
         %44 = OpLoad %uint %43
         %45 = OpIEqual %bool %42 %uint_0
         %46 = OpSelect %_ptr_StorageBuffer__arr_float_uint_3 %45 %33 %34
         %47 = OpAccessChain %_ptr_StorageBuffer_float %46 %44
         %48 = OpLoad %float %47
               OpStore %40 %48
               OpReturn
               OpFunctionEnd
```
Note the OpSelect at %46.  Its operands are pointers into Private, but its result is pointer to StorageBuffer.  That's invalid.   We need the initializers for %33 and %34 but even with VariablePointers we can't select between two different pointer-to-Private values.

--------

See also other cases from comments:
- https://github.com/google/clspv/issues/2#issuecomment-333942922
See workarounds:
- rewrite the loads as function calls, to isolate them from instcombine: https://github.com/google/clspv/issues/2#issuecomment-334167822
- rewrite the tables as two-dimensional: https://github.com/google/clspv/issues/2#issuecomment-334169212  (This is less general than wrapping in a function)

Another effective workaround is to use -O0 to disable the inst-combine optimization.   But that's rather drastic.

**Comment by dneto0** [comment]
A relatively nice compiler fix would be to emulate the "rewrite as function" workaround:

- Before running inst-combine, rewrite getelementptr over private storage class as a non-pure function
- Then run instcombine
- Then convert the getelementptr functions back into real GEPs.

The main complexity is in generating all the right types for the functions.

**Comment by dneto0** [comment]
A simpler variation is: rather than rewriting the getelementptr instruction, wrap the *result* with a call to an opaque function.   That's a lower complexity solution.

**Comment by dneto0** [comment]
Wrapping the *pointer* with a function call is not enough.   The instcombine pass will still generate code like this:

```
cond.true:                                        ; preds = %entry
  %arrayidx = getelementptr inbounds [3 x float], [3 x float] addrspace(2)* @kFirst, i32 0, i32 %i
  %wrapped0 = call float addrspace(2)* @wrap(float addrspace(2)* %arrayidx) #1
  br label %cond.end

cond.false:                                       ; preds = %entry
  %arrayidx1 = getelementptr inbounds [3 x float], [3 x float] addrspace(2)* @kSecond, i32 0, i32 %i
  %wrapped1 = call float addrspace(2)* @wrap(float addrspace(2)* %arrayidx1) #1
  br label %cond.end

cond.end:                                         ; preds = %cond.false, %cond.true
  %cond.in = phi float addrspace(2)* [ %wrapped0, %cond.true ], [ %wrapped1, %cond.false ]
  %cond = load float, float addrspace(2)* %cond.in, align 4
  store float %cond, float addrspace(1)* %A, align 4
  ret void
}
```

**Comment by dneto0** [comment]
Wrapping the result of the load is sufficient to have it survive instcombine without doing select between the pointers:

This input survives intact through instcombine.
```
define float @wrap(float %A) {
entry:
        ret float  %A;
}


; Function Attrs: nounwind
define spir_kernel void @foo(float addrspace(1)* %A, i32 %c, i32 %i) #0 !kernel_arg_addr_space !3 !kernel_arg_access_qual !4 !kernel_arg_type !5 !kernel_arg_base_type !5 !kernel_arg_type_qual !6 {
entry:
  %cmp = icmp eq i32 %c, 0
  br i1 %cmp, label %cond.true, label %cond.false

cond.true:                                        ; preds = %entry
  %arrayidx = getelementptr inbounds [3 x float], [3 x float] addrspace(2)* @kFirst, i32 0, i32 %i
  %0 = load float, float addrspace(2)* %arrayidx, align 4
 %wrapped0 = call float @wrap(float %0)
  br label %cond.end

cond.false:                                       ; preds = %entry
  %arrayidx1 = getelementptr inbounds [3 x float], [3 x float] addrspace(2)* @kSecond, i32 0, i32 %i
  %1 = load float, float addrspace(2)* %arrayidx1, align 4
 %wrapped1 = call float @wrap(float %1)
  br label %cond.end

cond.end:                                         ; preds = %cond.false, %cond.true
  %cond = phi float [ %wrapped0, %cond.true ], [ %wrapped1, %cond.false ]
  store float %cond, float addrspace(1)* %A, align 4
  ret void
}
```

**Comment by dneto0** [comment]
I have a pretty good prototype at https://github.com/dneto0/clspv/commits/hide-constant-loads 

It fails only one test, but that's expected.  It resolves the test case above.

**Comment by dneto0** [comment]
Fixed by #77
